File: processing/syntactic/expressions/slice_expr.py
from __future__ import annotations

from processing.codegen.code_builder import CodeBuilder
from processing.syntactic.expressions.expr import Expr, child_property


class SliceExpr(Expr):
    """ slice_expr:
        [infix] ':' [infix] [':' [infix]];
    """

    @child_property
    def start(self) -> Expr:
        pass

    @child_property
    def stop(self) -> Expr:
        pass

    @child_property
    def step(self) -> Expr:
        pass

    def __init__(
            self,
            parent: Expr = None,
            start: Expr = None,
            stop: Expr = None,
            step: Expr = None
    ):
        super().__init__(parent)
        self.start = start
        self.stop = stop
        self.step = step

    def to_axion(self, c: CodeBuilder):
        c += self.start, ':', self.stop
        if self.step is not None:
            c += ':', self.step

    def to_python(self, c: CodeBuilder):
        c += self.start, ':', self.stop, ':', self.step

File: Axion.Core/Processing/Syntactic/Expressions/SliceExpr.cs
using Axion.Core.Processing.CodeGen;

namespace Axion.Core.Processing.Syntactic.Expressions {
    /// <summary>
    ///     <c>
    ///         slice_expr:
    ///             [preglobal_expr] ':' [preglobal_expr] [':' [preglobal_expr]];
    ///     </c>
    /// </summary>
    public class SliceExpr : Expr {
        private Expr from;

        internal Expr From {
            get => from;
            set => SetNode(ref from, value);
        }

        private Expr to;

        public Expr To {
            get => to;
            set => SetNode(ref to, value);
        }

        private Expr step;

        public Expr Step {
            get => step;
            set => SetNode(ref step, value);
        }

        public SliceExpr(
            Expr parent = null,
            Expr from   = null,
            Expr to     = null,
            Expr step   = null
        ) : base(parent) {
            From = from;
            To   = to;
            Step = step;

            MarkPosition(from ?? to ?? step, step ?? to ?? from);
        }

        public override void ToAxion(CodeWriter c) {
            c.Write("[", From, ":", To, ":", Step, "]");
        }

        public override void ToCSharp(CodeWriter c) {
            c.Write(".Slice(", From, ", ", To, ", ", Step, ")");
        }

        public override void ToPython(CodeWriter c) {
            ToAxion(c);
        }
    }
}